static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.ts

```diff
@@ -1,38 +1,69 @@
+import { log } from '@guardian/libs';
 import { getUrlVars } from '../../../../lib/url';
 import { getAdvertById } from './get-advert-by-id';
 import { enableLazyLoad } from './lazy-load';
+import { memoizedFetchNonRefreshableLineItemIds } from './non-refreshable-line-items';
+import { shouldRefresh } from './should-refresh';
 
-const setSlotAdRefresh = (event: googletag.events.Event): void => {
+const setSlotAdRefresh = (
+	event: googletag.events.ImpressionViewableEvent,
+): void => {
 	const advert = getAdvertById(event.slot.getSlotElementId());
 	const viewabilityThresholdMs = 30000; // 30 seconds refresh
 
-	if (advert?.shouldRefresh) {
-		const onDocumentVisible = () => {
-			if (!document.hidden) {
-				document.removeEventListener(
-					'visibilitychange',
-					onDocumentVisible,
+	if (!advert) {
+		return;
+	}
+
+	// Asynchronously retrieve the non-refreshable line item ids
+	// Only do this if they haven't been attached to the page config
+	const { tests, page } = window.guardian.config;
+	if (
+		tests?.fetchNonRefreshableLineItemsVariant === 'variant' &&
+		!page.nonRefreshableLineItemIds
+	) {
+		// Call the memoized function so we only retrieve the value from the API once
+		void memoizedFetchNonRefreshableLineItemIds()
```

Comment:
We also have `async-call-merger` that does this memoization for network calls: https://github.dev/guardian/frontend/blob/3bb4ab3976ba02b28b035b8186dc9b52eacce054/static/src/javascripts/projects/common/modules/async-call-merger.ts

Reply:
This looks useful but I might have to spend some time figuring out how it works. Maybe for a later PR?